Updated ServiceCode for m3u8 links

Twoure opened the pull request:
_(08/11/2016)_

Did little testing so not sure if all links are now `m3u8`, but the few I did test were solely `m3u8` links.  Also cleaned up some of the `MetadataObjectForURL()` code.

Cheers

Twoure committed:
Updated ServiceCode for m3u8 changes

Twoure committed:
Updated version to 0.06

cksodhi commented:
I'm getting an error: Cannot load m3u8 crossdomain access denied

This shows up when I access something from the web interface.

Did you get the same error?

Twoure commented:
Oh, yup.  Didn't check the Plex Web client, only the Android client.  Normally you can disable the direct stream/play for Plex Web client to bypass the crossdomain error, but it doesn't seem to work for these streams.

~~If I recall correctly, there is a limited number of clients that can support direct eHLS streams _(which these are)_.  Most likely the channel's _Info.plist_ will need to be updated with the allowed clients.  More info can be found in [Comment_1164642](https://forums.plex.tv/discussion/comment/1164642/#Comment_1164642).~~ Channel host normal HLS not eHLS I guess ([ref](https://github.com/coder-alpha/Einthusan.bundle/pull/3#issuecomment-239301645)).

Twoure committed:
ServiceCode - removed 1hour cache time

cksodhi commented:
I'm still not able to get it to work .. not even on android! ... I'm wondering if this will work on Roku!

I will try when I get home tonight and let you know. 

Nonetheless, thank you so much for working on it and trying to fix it. 

Cheers!

ghost commented:
> If I recall correctly, there is a limited number of clients that can support direct eHLS streams (which these are). Most likely the channel's Info.plist will need to be updated with the allowed clients. More info can be found in Comment_1164642.

PMS _should_ be able to convert encrypted HTTP Live Stream video into another format. The Info.plist file doesn't contain compatibility information at all :(

ghost commented:
Looking at the m3u8 playlist files, the streams are normal HTTP Live Streams, not encrypted ones. There's no information about encryption keys in the playlists at all.

Twoure commented:
@piplongrun 

> PMS should be able to convert encrypted HTTP Live Stream video into another format. The Info.plist file doesn't contain compatibility information at all :(

Sorry meant to say, if the streams only work for certain clients, then the valid clients could be included withing the Info.plist file.

> Looking at the m3u8 playlist files, the streams are normal HTTP Live Streams, not encrypted ones. There's no information about encryption keys in the playlists at all.

Thanks for the insight.

Twoure committed:
ServiceCode - fixed user-agent error.

When direct stream/play is off the Plex transcoder was percent encoding the user-agent cause an HTTP 403 error with the stream because the stream has a hash tied with the user-agent.  Fixed by shortening the relevant user-agent within the stream request so percent encoding will not effect string.

Twoure commented:
Ok, figured out what the issue was.  When direct stream/play is disabled, it enables the Plex transcoder to handle the stream.  Currently there is an issue, where the transcoder is percent encoding the `user-agent` causing an HTTP 403 error within the stream.  This is because the stream is hashed by the request `user-agent`, and if it is different when calling the `ts` streams or `m3u8` playlist, then it fails to authenticate.  Luckily this channel accepted a shortened `user-agent` that contained no charterers that would end up being percent encoded for the stream URLs.

https://github.com/coder-alpha/Einthusan.bundle/pull/3/commits/7e356ff1327739c42de50e1fd6e66cf038b7f29e should fix the transcoding issue.

So, after downloading the latest patch, if the error **Cannot load M3U8: crossdomain access denied** shows up within a client, then try disabling the Plex client's direct play/stream.

---

Note:  I'm running the latest Plex-Pass PMS _v1.0.3_ on Ubuntu 14.04 64bit

Twoure commented:
@cksodhi 
After the latest push, does the pull request work for you now?

viveksupe commented:
Nope. Still not working.

cksodhi commented:
@Twoure the web client works great now. Android isnt working anymore, I get "Playback has stopped because the connection to the PMS has been lost. ..... " I think plex on Roku is also working.

This is really good @Twoure , It's great to know that this channel isnt abandoned yet. :) Thank you so much! Cheers!

ajaffarali commented:
I'm getting a "Channel is not responding" message when I click on the plugin and then Alphabetical or any other option

sahilcc7 commented:
0.06 is working for me on web after disabling direct play. It also works on Android after disabling direct play in the settings. Haven't tried Chromecast yet. 

EDIT: Chromecast not working.
